Design.md

```diff
@@ -0,0 +1,149 @@
+
+## Endpoints
+
+### Start
+
+Starts a new job by spawning a process from given executable path and arguments. Returns a unique job id (UUID v4), that is used to specify the target job for other endpoints. If the executable is not found or cannot be executed, immediately returns an error.
+
+No security checks are applied to the program and arguments. If the client wishes to remove some files or start a fork bomb, it's allowed to do so.
+
+### Stop
+
+Cancels a job by sending `SIGKILL`. Stop only returns after the process has been terminated, and returns the status of the job. If job was already terminated before calling stop, the old state is returned without attempting termination.
+
+Could be improved by sending `SIGTERM` shortly before `SIGKILL`, but I'm aiming for simplicity here. Also separating soft and hard kills in the API level might be useful in some situations.
+
+### Status
+
+Returns job status, i.e. is it running, and the status code if the job has completed. If the job has been terminated with a signal, that is reported instead.
+
+```proto
+message JobStatus {
+    enum Status {
+        RUNNING = 0;    // Still running (value is zero)
+        COMPLETED = 1;  // Completed normally (value is return code)
+        SIGNAL = 2;     // Terminated by a signal (value is signal number)
+        MISSING = 3;    // No such process (or access denied) (value is zero)
+    }
+    Status status = 1;
+    int32 value = 2;
+}
+```
+
+### Output
+
+Streams output of a job in binary blobs. Each blob is tagged to be either from stdout or stderr. Stream is automatically closed when the process completes and all output has been streamed. All calls to output stream the whole output history from the moment the process was started.
+
+#### Internals
+
+There will be two async-tasks reading the output of a process, one for stdout and one for stderr. For both tasks, a separate buffer of output history is maintained, along with a boolean marking process completion. This state is protected by a [`tokio::sync::RwLock`](https://docs.rs/tokio/1.6.1/tokio/sync/struct.RwLock.html). In addition, there will be a [`tokio::sync::Notify`](https://docs.rs/tokio/1.6.1/tokio/sync/struct.Notify.html). Every time the process writes more output, the listerners are woken up through the `Notify`.
+
+Each call of `Output` spawns async-tasks for stdio and stdout. They read the output buffer until the end. Then it checks if the process is completed (from the field). If yes, then the connection to client is closed to mark procress completion. Otherwise, it waits until the output reader task notifies it that new data is available, and then repeats the above process.
```

Comment:
```suggestion
Each call of `Output` spawns async-tasks for stdio and stdout. They read the output buffer until the end. Then it checks if the process is completed (from the field). If yes, then the connection to client is closed to mark process completion. Otherwise, it waits until the output reader task notifies it that new data is available, and then repeats the above process.
```